design/one-pager-resource-connectivity-mvp.md

```diff
@@ -0,0 +1,653 @@
+        acceleratorType: large
+      diskType: pd-ssd
+      minCpuPlatform: Intel Haswell
+      taints:
+        - key: color
+          value: purple
+          effect: NO_SCHEDULE
+    initialNodeCount: 3
+    version: 1.14
+    autoscaling:
+      enabled: true
+      minNodeCount: 3
+      maxNodeCount: 9
+    management:
+      autoUpgrade: true
+      autoRepair: true
+    maxPodsConstraint:
+      maxPodsPerNode: 254
+  locations:
+  - us-central1-a
+  enableKubernetesAlpha: false
+  resourceLabels:
+    cool: very
+  legacyAbac:
+    enabled: false
+  networkPolicy:
+    provider: CALICO
+    enabled: true
+  ipAllocationPolicy:
+    useIpAliases: true
+    createSubnetwork: false
+    subnetworkName: mycoolsubnetwork
+    clusterSecondaryRangeName: mycoolpodsrange
+    servicesSecondaryRangeName: mycoolservicesrange
+    clusterIpv4CidrBlock: 10.0.0.0/8
+    nodeIpv4CidrBlock: 172.16.1.0/24
+    servicesIpv4CidrBlock: 192.168.0.0/24
+    tpuIpv4CidrBlock: 172.16.2.0/24
+  masterAuthorizedNetworksConfig:
+    enabled: true
+    cidrBlocks:
+      displayName: coolcidr
+      cidrBlock: 93.80.0.0/16
+  maintenancePolicy: object (MaintenancePolicy)
+    window:
+      dailyMaintenanceWindow:
+        startTime: 01:30
+  defaultMaxPodsConstraint:
+    maxPodsPerNode: 254
+  resourceUsageExportConfig:
+    bigqueryDestination:
+      datasetId: mycooldataset
+    enableNetworkEgressMetering: true
+    consumptionMeteringConfig:
+      enabled: true
+  privateClusterConfig:
+    enablePrivateNodes: true
+    enablePrivateEndpoint: true
+    masterIpv4CidrBlock: 172.16.0.0/16
+  initialClusterVersion: 1.13
+  enableTpu: false
+```
+
+## Proposal
+
+Minimum viable support for resource connectivity can be enabled by ensuring
+three things.
+
+* All external resources required to configure connectivity have corresponding
+  Crossplane managed resources, such that the cloud administrator does not need
+  to leave Crossplane to configure them. For example, if a `KubernetesCluster`
+  can only connect to a `MySQLInstance` in the same VPC network then Crossplane
+  must be able to represent said VPC network as a managed resource.
+* All external resource settings required to configure connectivity must be
+  exposed in the configuration of their associated managed resource.
+* All managed resource settings required to configure connectivity must be
+  exposed in the configuration of their associated resource class.
+
+With the above in place resource connectivity between a `KubernetesCluster` and
+a `MySQLInstance` can be configured roughly as follows:
+
+1. The cloud administrator creates any managed resources necessary to connect
+   a Kubernetes cluster to a MySQL database in their cloud of choice, for
+   example they create a VPC network for both external resources to live in.
+1. The cloud administrator creates resource classes to be used when creating
+   `KubernetesCluster` and `MySQLInstance` resource claims against their cloud
```

Comment:
In the world of "single" resource classes per type, I think this is manageable. However, how will this address the cases where the environment has multiple `KubernetesCluster` classes with different network configuration and multiple `MySQLInstance` classes, also with different network configurations?

On the separate note, how would this work if an application needs to "consume" two different resources (doesn't have to be even the same resource type) that belong to two different network configurations? As an example: "redis" deployed using prviate "networkA" and "cloudsql" is deployed using private "networkB", and the application needs to consume both.

Reply:
We will need to solve such problems, but they're out of scope for this initial (very) MVP design.